internal/db/schema/postgres/postgres.go

```diff
@@ -0,0 +1,345 @@
+// Original Work
+// Copyright (c) 2016 Matthias Kadenbach
+// https://github.com/mattes/migrate
+//
+// Modified Work
+// Copyright (c) 2018 Dale Hui
+// https://github.com/golang-migrate/migrate
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/go-multierror"
+	"github.com/lib/pq"
+)
+
+// schemaAccessLockId is a Lock key used to ensure a single boundary binary is operating
+// on a postgres server at a time.  The value has no meaning and was picked randomly.
+const (
+	schemaAccessLockId int64 = 3865661975
+	nilVersion               = -1
+)
+
+var defaultMigrationsTable = "boundary_schema_version"
+
+// Postgres is a driver usable by a boundary schema manager.
+type Postgres struct {
+	// Locking and unlocking need to use the same connection
+	conn *sql.Conn
+	db   *sql.DB
+}
+
+// New returns a postgres pointer with the provided db verified as
+// connectable and a version table being initialized.
+func New(ctx context.Context, instance *sql.DB) (*Postgres, error) {
+	const op = "postgres.New"
+	if err := instance.PingContext(ctx); err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+	conn, err := instance.Conn(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+
+	px := &Postgres{
+		conn: conn,
+		db:   instance,
+	}
+
+	if err := px.ensureVersionTable(ctx); err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+
+	return px, nil
+}
+
+// TrySharedLock attempts to capture a shared lock. If it is not successful it returns an error.
+// https://www.postgresql.org/docs/9.6/static/explicit-locking.html#ADVISORY-LOCKS
+func (p *Postgres) TrySharedLock(ctx context.Context) error {
+	const op = "postgres.(Postgres).TrySharedLock"
+	r := p.conn.QueryRowContext(ctx, "SELECT pg_try_advisory_lock_shared($1)", schemaAccessLockId)
```

Comment:
Previous comments are not showing up for some reason. 

[Me](https://github.com/hashicorp/boundary/pull/842#discussion_r558595089):
> Why is there no equivalent unlock method for this (`select pg_advisory_unlock_shared($1)`)?

[Todd](https://github.com/hashicorp/boundary/pull/842#discussion_r558631304):
> Just because we do not have a use for it yet.

Shouldn't we be releasing the lock by calling `select pg_advisory_unlock_shared($1)`? Or are we just letting Postgres do the cleanup when the connection closes? If it is the latter, are there any dangers on relying on that?